#0: Clean up some include paths around RTOptions and remove direct checks on TT_METAL_SIMULATOR env var

tt-dma opened the pull request:
Follow-up to https://github.com/tenstorrent/tt-metal/pull/16375, cleans up the uses of TT_METAL_SIMULATOR and moves that setting into RTOptions. Required a bit of header file cleanup.

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/13107473655
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

tt-dma committed:
#0: Clean up some include paths around RTOptions and remove direct
checks on TT_METAL_SIMULATOR env var

tt-dma commented on tt_metal/api/tt-metalium/get_platform_architecture.hpp:
```diff
@@ -50,7 +51,7 @@ namespace tt::tt_metal {
 inline tt::ARCH get_platform_architecture() {
     auto arch = tt::ARCH::Invalid;
     if (std::getenv("TT_METAL_SIMULATOR")) {
```
For some reason fixing this line specifically gives a linker error for `fabric_unit_tests` executable only (everything else builds fine):
```
[551/1013] Linking CXX executable test/tt_metal/tt_fabric/fabric_unit_tests
FAILED: test/tt_metal/tt_fabric/fabric_unit_tests tests/tt_metal/tt_fabric/fabric_unit_tests[1]_tests.cmake /home/davidma/tt-metal/build_RelWithDebInfo/tests/tt_metal/tt_fabric/fabric_unit_tests[1]_tests.cmake
: && /usr/bin/clang++-17 -stdlib=libc++ -O3 -g -DDEBUG -lc++ -lc++abi tests/tt_metal/tt_fabric/CMakeFiles/fabric_unit_tests.dir/fabric_router/test_routing_tables.cpp.o -o test/tt_metal/tt_fabric/fabric_unit_tests  -Wl,-rpath,/home/davidma/tt-metal/build_RelWithDebInfo/tt_fabric:/home/davidma/tt-metal/build_RelWithDebInfo/tt_metal:/home/davidma/tt-metal/build_RelWithDebInfo/tt_metal/third_party/umd/device:/home/davidma/tt-metal/build_RelWithDebInfo/lib  tt_fabric/libtt_fabric.so  tt_metal/libtt_metal.so  tt_metal/third_party/umd/device/libdevice.so  _deps/flatbuffers-build/libflatbuffers.a  -lpthread  lib/libgmock_main.so.1.13.0  lib/libgmock.so.1.13.0  lib/libgtest.so.1.13.0  -lpthread && cd /home/davidma/tt-metal/build_RelWithDebInfo/tests/tt_metal/tt_fabric && /usr/bin/cmake -D TEST_TARGET=fabric_unit_tests -D TEST_EXECUTABLE=/home/davidma/tt-metal/build_RelWithDebInfo/test/tt_metal/tt_fabric/fabric_unit_tests -D TEST_EXECUTOR= -D TEST_WORKING_DIR=/home/davidma/tt-metal/build_RelWithDebInfo/tests/tt_metal/tt_fabric -D TEST_EXTRA_ARGS= -D TEST_PROPERTIES= -D TEST_PREFIX= -D TEST_SUFFIX= -D TEST_FILTER= -D NO_PRETTY_TYPES=FALSE -D NO_PRETTY_VALUES=FALSE -D TEST_LIST=fabric_unit_tests_TESTS -D CTEST_FILE=/home/davidma/tt-metal/build_RelWithDebInfo/tests/tt_metal/tt_fabric/fabric_unit_tests[1]_tests.cmake -D TEST_DISCOVERY_TIMEOUT=5 -D TEST_DISCOVERY_EXTRA_ARGS= -D TEST_XML_OUTPUT_DIR= -P /usr/share/cmake-3.31/Modules/GoogleTestAddTests.cmake
libc++abi: __cxa_guard_acquire detected recursive initialization: do you have a function-local static variable whose initialization depends on that function?
CMake Error at /usr/share/cmake-3.31/Modules/GoogleTestAddTests.cmake:132 (message):
  Error running test executable.
```

I haven't been able to make heads or tails of what's going on here, have any compiler pros seen this error before?

ppenzin replied:
This is coming from a constructor guard for static instances, it tries to detect if it has been called recursively, which it can't handle ([patch](https://reviews.llvm.org/D148630)).

ppenzin replied:
Function local static variable downstream from a call to static constructor?

tt-dma replied:
Issue ends up being a circular dependency with the RTOptions static constructor, see second commit for details.

tt-vjovanovic commented on tt_metal/api/tt-metalium/rtoptions.hpp:
```diff
@@ -133,6 +133,9 @@ class RunTimeOptions {
 
     bool skip_deleting_built_cache = false;
 
+    bool simulator_enabled = false;
```
In my changes, I tried to have `std::optional<tt_SimulationDeviceInit> simulatior_initialization` member in this class and initialize it in constructor of `RunTimeOptions` (since initialization of that class actually parses yaml file, etc.) and on all other places just use that instance.
Function `bool get_simulator_enabled()` would return `simulatior_initialization.has_value()`...
I wasn't sure if `RunTimeOptions` was supposed to have `tt_SimulationDeviceInit` instance in it, but sounded like a good trade of :)

tt-dma replied:
I think the plan for RTOptions is to stick to just holding env var data, so I'll leave it as is for now

omilyutin-tt commented on tt_metal/common/core_descriptor.cpp:
```diff
@@ -3,11 +3,57 @@
+            case tt::ARCH::GRAYSKULL: return core_desc_dir + "grayskull_versim_1x1_arch.yaml";
+            case tt::ARCH::WORMHOLE_B0: return core_desc_dir + "wormhole_b0_versim_1x1_arch.yaml";
+            case tt::ARCH::BLACKHOLE: return core_desc_dir + "blackhole_simulation_1x2_arch.yaml";
+            default: throw std::runtime_error("Unsupported device arch");
```
Just FYI - we now have the compiler configured to error out if there is an unhandled `case:` and no `default:` branch. Doing so helps developers who add new enum values to surface all places in the codebase that enumerate the them - so that you know which places you need to modify. Consider removing `default:` here.

tt-dma replied:
Didn't know that, thanks. Will remove the defaults here

tt-dma committed:
#0: Fix RTOptions()-DispatchCoreConfig() circular dependency

Issue was that RTOptions had a DispatchCoreConfig member, which requires
some existing RTOptions settings to initialize. Break the loop by
constructing a DispatchCoreConfig only when we need it (used for
gtesting).
